### PR TITLE
WIP: Teach Indexify about Raft

### PR DESCRIPTION
**Pull request by jackbackes**
This is majorly a work in progress, as I'm learning the details of the Raft algorithm as I go :-)
Here is the status:

Done:
1. Add in-memory interface for OpenRaft
2. Add Raft management and communication APIs
3. Add Raft RPC protocol layer
4. Integrate coordinator state into Raft state machine
5. Integrate coordinator api layer

TODO:

1. Test
2. Fix merge conflicts, Clean up, document, make pretty

Future PR:

1. Sled integration

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -9,17 +9,19 @@ members = [".", "migration"]
 
 [dependencies]
 anyhow = { version = "^1" }
-async-trait = "*"
+async-trait = "0.1.74"
```
Comment (jackbackes) [inline Cargo.toml:12]: note to self: revert this.

File: Cargo.toml [hunk 2]
```diff
@@ -9,17 +9,19 @@ members = [".", "migration"]
...
 axum = { version = "^0", features = ["multipart"] }
 axum-macros = { version = "0.3.7" }
 axum-otel-metrics = "0.6"
 axum-tracing-opentelemetry = "0.14.1"
+byteorder = "1.5.0"
```
Comment (jackbackes) [inline Cargo.toml:17]: note to self: no longer needed

File: Cargo.toml [hunk 3]
```diff
@@ -52,6 +55,7 @@ serde_json = { version = "^1", features = ["std"] }
 smart-default = { version = "^0" }
 strum = { version = "^0", features = ["derive"] }
 strum_macros = { version = "^0" }
+tempfile = "3.8.0"
```
Comment (jackbackes) [inline Cargo.toml:58]: note to self: revert this

File: src/bin/test_raft_setup.rs [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+		.route("/raft/append", axum::routing::post(raft_api::append))
+		.route("/raft/snapshot", axum::routing::post(raft_api::snapshot))
+		.route("/raft/vote", axum::routing::post(raft_api::vote))
+		// admin API
+		.route("/management/init", axum::routing::post(management::init))
+		.route("/management/add_learner", axum::routing::post(management::add_learner))
+		.route("/management/change_membership", axum::routing::post(management::change_membership))
+		.route("/management/metrics", axum::routing::get(management::metrics))
+		// application API
+		// TODO: ADD THE COORDINATOR API HANDLERS HERE
```
Comment (jackbackes) [inline src/bin/test_raft_setup.rs:41]: Note: this is where Coordinator API handlers will go.

File: src/coordinator.rs [hunk 5]
```diff
@@ -27,27 +28,34 @@ use std::{
 
 #[derive(Debug)]
 pub struct Coordinator {
+    data: CoordinatorData,
```
Comment (jackbackes) [inline src/coordinator.rs:31]: The state machine needs serializable data so extracted the data from the repo metadata.

File: src/raft/api.rs [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+ *
+ *
+ *  - `POST - /record_executor
+ *  - `POST - /sync_executor
+ *  - `POST - /create_work
+ *  - `POST - /embed_query
+ */
+pub struct Api {}
+
+ // TODO: IMPLEMENT THIS
```
Comment (jackbackes) [inline src/raft/api.rs:12]: This is WIP

File: src/raft/memstore.rs [hunk 7]
```diff
@@ -0,0 +1,427 @@
...
+use openraft::StoredMembership;
+use openraft::Vote;
+use serde::Deserialize;
+use serde::Serialize;
+use tokio::sync::RwLock;
+
+use crate::coordinator::CoordinatorData;
+
+/// The application data request type which the `MemStore` works with.
+/// TODO: update to handle the coordinator data
```
Comment (jackbackes) [inline src/raft/memstore.rs:38]: Todo: update to handle coordinator <-> Raft communication

File: src/raft/raft_api.rs [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+use super::coordinator_config::CoordinatorRaftApp as App;
+use super::memstore::MemNodeId as NodeId;
+use super::memstore::Config as TypeConfig;
+
+// --- Raft communication
+
+#[axum_macros::debug_handler]
+pub async fn vote(app: Extension<App>, Json(req): Json<VoteRequest<NodeId>>) -> impl IntoResponse {
+    let res = app.raft.vote(req).await;
+    AxumJson(res)
```
Comment (diptanu) [inline src/raft/raft_api.rs:12]: write the actual type in the output?

File: src/raft/coordinator_config.rs [hunk 9]
```diff
@@ -0,0 +1,37 @@
+use openraft::Raft;
+use crate::server_config::CoordinatorConfig;
+
+use super::{memstore::{MemNodeId, Config as CoordinatorRaftConfig, MemStore}, network::IndexifyRaftNetwork};
+use std::sync::Arc;
+
+pub type CoordinatorRaft = Raft<CoordinatorRaftConfig, IndexifyRaftNetwork, Arc<MemStore>>;
+
+#[derive(Clone)]
+pub struct CoordinatorRaftApp {
```
Comment (diptanu) [inline src/raft/coordinator_config.rs:10]: Is this the application state ?

File: src/raft/coordinator_node.rs [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+
+impl RaftCoordinatorNode {
+    pub async fn new(id: MemNodeId, addr: String, raft_config: openraft::Config, coordinator_config: CoordinatorConfig) -> Result<Self, anyhow::Error> {
+        let coordinator = CoordinatorRaftApp::new(id, addr, raft_config, coordinator_config).await?;
+        let addr: SocketAddr = coordinator.addr.parse()?;
+        info!("coordinator listening on: {}", addr.to_string());
+        Ok(Self { addr, coordinator: Arc::new(coordinator) })
+    }
+
+    pub async fn run(&self) -> Result<(), anyhow::Error> {
```
Comment (diptanu) [inline src/raft/coordinator_node.rs:25]: Merge the routes here with coordinator routers with a prefix `/raft/`

File: src/raft/api.rs [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+use crate::internal_api::ListExecutors;
+use crate::internal_api::SyncExecutor;
+use crate::internal_api::SyncWorkerResponse;
+use crate::raft::memstore::Request;
+
+use super::coordinator_config::CoordinatorRaftApp as App;
+
+ // TODO: maybe combine with non-raft API so there aren't two separate handlers
+ #[axum_macros::debug_handler]
+ pub(super) async fn sync_executor(
```
Comment (diptanu) [inline src/raft/api.rs:18]: We can remove the code which does all this in the coordinator.rs, no need to maintain two different code paths.

File: src/raft/api.rs [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+		Ok(res) => res.data.into(),
+		Err(e) => {
+			return Err(IndexifyAPIError::new(StatusCode::INTERNAL_SERVER_ERROR, format!("Error writing to Raft: {}", e)))
+		}
+	};
+	Ok(Json(response))
+}
+
+#[axum_macros::debug_handler]
+pub(super) async fn embed_query(
```
Comment (diptanu) [inline src/raft/api.rs:46]: this will go away when you rebase with master.

File: src/raft/memstore.rs [hunk 13]
```diff
@@ -0,0 +1,668 @@
...
+use crate::persistence::Repository;
+use crate::persistence::{ExtractorConfig, Work};
+
+use crate::coordinator::CoordinatorData;
+use crate::server_config::CoordinatorConfig;
+
+// TODO: these could probably just wrap the underlying work types rather than duplicating the fields
+/// The application data request type which the `MemStore` works with.
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub enum Request {
```
Comment (diptanu) [inline src/raft/memstore.rs:51]: Yeah no need to duplicate all the enums below

**Review by diptanu — CHANGES_REQUESTED** [review]
Left some comments, looks like there is a bunch of duplication between various existing apis and what we are introducing to support raft. Lets use the server we already have to expose raft endpoints. In terms of the apis around requests, I think using the current types would be better than duplicating them. The Raft config should be merged with server config.

**Comment by diptanu** [comment]
Raft is already in!

**Pull request closed without merging**